Vimc 7111: initial version of package vacsce

xiangli313 opened the pull request:
Dear all,

This PR is related to the initial set-up of package names ```vacsce```. 

This package stores coverage projection rules. We intend to wrap them up as a package usable by both internal and external users.

Thanks!

xiangli313 committed:
initiation

xiangli313 committed:
readme

xiangli313 committed:
comments

xiangli313 committed:
progress

xiangli313 committed:
upload

xiangli313 committed:
save progress

xiangli313 committed:
upload

xiangli313 committed:
progress upload

xiangli313 committed:
upload

xiangli313 committed:
updates

xiangli313 committed:
upload

xiangli313 committed:
upload

xiangli313 committed:
update man

xiangli313 committed:
upload

xiangli313 committed:
update tests

xiangli313 committed:
move

xiangli313 committed:
updates

xiangli313 committed:
updates: add age and gender to routine rules

xiangli313 committed:
gitignore

xiangli313 committed:
upload html

richfitz commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+  } else {
+    input$his <- his
+    input$introduction <- input$params$introduction
+    print("No historical routine data identified.")
```
prefer `message()` to `print()`

Is continuing here what you really want to do?

richfitz commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+  ## run projection rules
+  ## scale routine coverage back by proportion risk
+  his <- his %>%
+    mutate(coverage = ifelse(activity_type == "routine", coverage/input$params$proportion_risk, coverage))
```
I think that `dplyr::case_when` is the preferred way of doing this, `ifelse` has lots of problems. @kgaythorpe would know better than me though

richfitz commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+vac_sce <- function(input){
+  ## log scenario generation messages.
+  input <- input_check(input)
+  his <- input$his
```
personally, I would expand these sorts of names to `history` and `future` because they're not much longer but easier to work with

xiangli313 replied:
Done.

richfitz commented on R/util.R:
```diff
@@ -0,0 +1,57 @@
+## ia2030 non-linear function
+IA2030_projection <- function(t0, c0, T, cT){
```
I'd be careful not to use `T` here for the name, it looks like `TRUE` (see the syntax highlighting here pushing you this way). More descriptive names in general would help if you needed someone like me to help with the function in the future

kgaythorpe replied:
agreed, also minor comment- could use lower case name for function for consistency- would also add citation to the ia2030 paper (Carter et al. 2023 https://www.sciencedirect.com/science/article/pii/S0264410X2300854X)

xiangli313 replied:
renamed to year_base, coverage_base, year_target, and coverage_target.

richfitz commented on R/test_tool.R:
```diff
@@ -0,0 +1,106 @@
+# rm(list = ls())
```
This file should be deleted.

richfitz commented on R/projection_rules.R:
```diff
@@ -0,0 +1,275 @@
+
+  ## project future campaigns from y_base to year_to
+  i <- y_base
+  t <- NULL
```
```suggestion
  year <- NULL
```

(and then below when used) so that later you go


```
  t <- data.frame(year = year,
                  coverage = sia_level,
                  age_from = age_from,
                  age_to = age_to,
                  gender = gender)
```

which will be easier to work with in a year's time

xiangli313 replied:
Done.

richfitz commented on R/projection_rules.R:
```diff
@@ -0,0 +1,275 @@
+  d <- d[is.numeric(d$year) & d$year > 0,]
+  if (nrow(d) == 0){
+    ## routine introduction
+    cov[1] <- vaccine_x_level*intro_level # intro at 33% of target vaccine
```
```suggestion
    cov[1] <- vaccine_x_level*intro_level # intro at 33% of target vaccine by default (see args)
```

richfitz commented on R/projection_rules.R:
```diff
@@ -0,0 +1,275 @@
+keep_levels <- function(d, year_from, year_to, level, gender = NULL, age_from = NULL, age_to = NULL){
+  t <- d[1, ] %>%
+    dplyr::select(-year) %>%
+    merge(data.frame(year = year_from:year_to)) %>%
```
I believe (and again would defer to @kgaythorpe) that one of the dplyr join functions here is preferable

kgaythorpe replied:
agreed

xiangli313 replied:
Done.

richfitz commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+  # hpv: catch-up + future routine intro
+  # mena: catch-up + mini-catch-up + future routine intro
+  # anonymous region
+  t <- data.frame(region = c(324, 324, 324, 50, 50, 108, 108),
```
Some explanation as to the origin of these numbers is essential

kgaythorpe replied:
agreed though as this is example data, these numbers could just be randomly generated

xiangli313 replied:
These are country.nid in montagu. Added notes for this.

richfitz commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+        func <- paste0(names(r[j]), sub("list\\(", "(d, ", paste(r[j])))
+        print(func)
+        d <- eval(parse(text = func))
+      }
```
I am not at all sure what this is trying to do but:

* the print statement suggests it's fairly tricky and took a while to work out :)
* there is probably a better way of doing it that will be easier to understand and more robust

Can you try and describe what is going on, and perhaps I can come up with some easier way

richfitz commented on dettl_config.yml:
```diff
@@ -0,0 +1,40 @@
+db:
```
why is there a dettl config here? that should be only in montagu-imports surely?

xiangli313 replied:
It was only used while extracting sample data for developing the package. Removed now.

richfitz commented on tests/testthat/test-vacsce-generation.R:
Empty file - delete or write some tests :)

xiangli313 replied:
Deleted.

richfitz commented on vignettes/Using vacsce for vaccination coverage scenario generation.html:
Delete generated files

xiangli313 replied:
Done.

richfitz commented on DESCRIPTION:
```diff
@@ -0,0 +1,39 @@
+Package: vacsce
```
Can you enable some standard github actions please:

```
usethis::use_github_action("pkgdown")
usethis::use_github_action("test-coverage")
usethis::use_github_action_check_standard()
```

then push the result up to github (will be a .github) directory. Most likely they will fail, but do so informatively

xiangli313 replied:
Done.

richfitz commented on NAMESPACE:
```diff
@@ -0,0 +1,2 @@
+# Generated by roxygen2: do not edit by hand
```
Your package exports no functions, which means it has no interface. Add `#' @export` tags to the functions you want to export and redocument

xiangli313 replied:
Done.

richfitz commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+.Rproj.user
```
Please add a basic README.md at the top level

xiangli313 replied:
There is a README.md in vacsce/ directory. Is that what you mean?

richfitz reviewed (changes requested):
Hi Xiang - some general comments here for you to consider. I'm sure Katy will have comments on the data processing code. I presume this has come from some previous report or import that you've consolidated?

Some of the comments are only written once but apply throughout (comments on paste, ifelse, etc). The section with the eval(parse()) is worth thinking about carefully.

Happy to loop back on this next week, but there's a chance I may be away

kgaythorpe commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+  if (nrow(his[his$activity_type != "campaign", ]) > 0){
+    s <- his %>% group_by(vaccine, activity_type) %>%
+      dplyr::filter(activity_type != "campaign") %>%
+      summarise(year_intro_src = min(year), .groups = "keep") %>%
```
also a dplyr function, summarise

xiangli313 replied:
Done.

kgaythorpe commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+  his <- his %>%
+    mutate(coverage = ifelse(activity_type == "routine", coverage/input$params$proportion_risk, coverage))
+  fut <- fut %>%
+    mutate(coverage = ifelse(activity_type == "routine", coverage/input$params$proportion_risk, coverage))
```
not sure if captured later but does this protect against dividing by 0

xiangli313 replied:
There is this code above - stopifnot(proportion_risk <= 1 & proportion_risk > 0) # proportion of at risk population. This makes sure proportion_risk is meaningfully within (0, 1].

kgaythorpe commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+      dat <- bind_rows(dat, d %>% bind_cols(input$introduction[i, c("vaccine", "activity_type")]))
+    } else if(is.null(r)){
+      print("No projection. Binding data from source coverage.")
+      dat <- bind_rows(dat, d0, d1)
```
dplyr::bind_rows

xiangli313 replied:
Done.

kgaythorpe commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+        func <- gsub('\"', "'", func, fixed = TRUE)
+        print(func)
+        dat <- eval(parse(text = func)) %>%
+          merge(input$introduction[i, c("vaccine", "activity_type")]) %>%
```
would stick to dplyr consistently rather than merge- fewer moving parts for things going wrong due to packages

xiangli313 replied:
Done.

kgaythorpe commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+                  disease =  c("Measles", "Measles", "Measles", "HPV", "HPV", "MenA", "MenA"),
+                  vaccine = c("MCV1", "MCV2", "Measles", "HPV", "HPV", "MenA", "MenA"),
+                  activity_type = c("routine", "routine", "campaign", "campaign", "routine", "campaign", "routine"))
+  d <- DBI::dbGetQuery(con,
```
is this just to generate data that will be included in the package or is this something that is going to be a problem for users without access to the database

xiangli313 replied:
This is just to generate example data for people who use this package. It is run only once. Using the package does not need access to the database.

kgaythorpe commented on R/general.R:
```diff
@@ -0,0 +1,213 @@
+                             "AND coverage > 0",
+                             "AND year <= 2030"),
+                       list(touch)) %>%
+    right_join(t, by = c("region", "vaccine", "activity_type")) %>%
```
more dplyr- mutate, right_join

xiangli313 replied:
Done.

kgaythorpe commented on R/projection_rules.R:
```diff
@@ -0,0 +1,275 @@
+#' @param age_from age_from
+#' @param age_to age_to
+incremental <- function(d, year_from, year_to, step, cap = 0.95, gender = NULL, age_from = NULL, age_to = NULL){
+  y <- max(d$year)
```
could this be renamed max_y for easier reading

xiangli313 replied:
Done.

kgaythorpe commented on R/projection_rules.R:
```diff
@@ -0,0 +1,275 @@
+#' @param age_to age_to
+incremental <- function(d, year_from, year_to, step, cap = 0.95, gender = NULL, age_from = NULL, age_to = NULL){
+  y <- max(d$year)
+  t <- d[d$year == y, ] %>%
```
would rename t for easier reading

xiangli313 replied:
Added a note : "this is just to grab one row in source data for information like region, vaccine, activity_type, age groups, etc."

kgaythorpe commented on R/util.R:
```diff
@@ -0,0 +1,57 @@
+}
+
+## extract source coverage for vimc user
+vimc_historical_data <- function(con, year_cur, coverage_src_his, coverage_src_fut){
```
worth having this not export or adding a message ("this is for VIMC internal use only")

xiangli313 replied:
This is removed. It was the initial function extracting example data.

kgaythorpe commented on vignettes/vignette.Rmd:
```diff
@@ -0,0 +1,516 @@
+
+## 1. Introduction
+
+This vignette describes how to use **vacsce** to generate vaccination coverage scenarios. **vacsce** provides a set of coverage projection rules. Its initial objective was to document commonly used coverage projection rules and work as a tool for VIMC's model runs, interim-update and related analysis. As VIMC continues expanding its collaboration with greater public health community, we further p ...
```
add something like "vasce stands for vaccination scenarios" so people know how to say it

xiangli313 replied:
Done.

kgaythorpe commented on vignettes/vignette.Rmd:
```diff
@@ -0,0 +1,516 @@
+
+We give an example where the rule is used for routine introduction. In this example, there is no historical MCV2 routine introduction, which indicats a future routine introduction. **catch_up_with_x** projects routine introduction in year `r year_cur + 3`. Coverage level increases non-linearly and reaches 85% in year `r year_cur + 6`. 85% represents coverage level (or up-/down-scaled level) of a  ...
+
+In the case where routine introduction has had happened. The projection rule will **non-linearly scale-up** to the target **vaccine_x_level** from **year_from-1**, hence one needs to make sure there is a coverage value in **year_from -1**. A message will show up saying that parameter **intro_level** is not applicable and not used. (N.B. **non-linearly scale-up** refers to the next generation rule ...
```
*has happened

xiangli313 replied:
I have a lot of double ** to make bold characters in html output.

richfitz replied:
Katy means that this should say

```suggestion
In the case where routine introduction has happened. The projection rule will **non-linearly scale-up** to the target **vaccine_x_level** from **year_from-1**, hence one needs to make sure there is a coverage value in **year_from -1**. A message will show up saying that parameter **intro_level** is not applicable and not used. (N.B. **non-linearly scale-up** refers to the next generation rule to be introduced.)
```

the `*` she uses there just indicates a correction, not the bolding

kgaythorpe commented on vignettes/vignette.Rmd:
````diff
@@ -0,0 +1,516 @@
+```
+
+**EXAMPLE 1**
+Example one works as a null run of **vacsce**. Without specification of projection rules, the package simply bind any historical and future coverage sources. In practice, a null run may be not useful at all for external users. However, for VIMC users, a null run works like combining historical and future from two data sources (e.g. touchstones).]
````
*will simply

kgaythorpe replied:
However, for VIMC users, a null run works by combining historical and future from two data sources (e.g. touchstones).]

xiangli313 replied:
Done.

kgaythorpe commented on vignettes/vignette.Rmd:
```diff
@@ -0,0 +1,516 @@
+
+**EXAMPLE 4**
+
+Comparing with example 3, example 4 have an additional campaign projection rule which projects future recurrent follow-up SIAs. Since campaign projection is dependent on routine projection, **vacsce** will prepare your input, and project routine and campaigns sequentially. You do not need to worry about the ordering of vaccine delivery in you input set-up. However, you need to make sure that the  ...
```
I would cut the"you do not need to worry" bit- as it sounds contradictory to the following sentence

xiangli313 replied:
Done.

xiangli313 committed:
remove non-useful file; prefer message to print

xiangli313 committed:
case_when

xiangli313 committed:
tidy case_when, fix vignette by adding proportion_risk to cholera example

xiangli313 committed:
standard github actions

xiangli313 committed:
export functions

xiangli313 committed:
dplyr

xiangli313 committed:
meaningful objects

xiangli313 committed:
comment on country nid, more dplyr

xiangli313 committed:
Update R/projection_rules.R

Co-authored-by: Rich FitzJohn <rich.fitzjohn@gmail.com>

xiangli313 committed:
dplyr

xiangli313 committed:
Merge branch 'vimc-7111' of github.com:vimc/vacsce into vimc-7111

xiangli313 committed:
meaningful objects

xiangli313 committed:
lower case for function

xiangli313 committed:
remove dettl config

xiangli313 committed:
dplyr

xiangli313 committed:
vignette

xiangli313 committed:
remove empty file, update readme

xiangli313 commented:
Hi Rich and Katy,

I think I have addressed most of your comments. 

Rich. The section with `eval(parse())` is one that I need your help with. The aim is to evaluate function and corresponding parameters from a string. So that a user can specify input in a standardise way, and let the package handle the rest. An example might be helpful here.
A user defines projection rules in the following set-up
```
rules = list(
          catch_up_with_x = list(year_from = 2027, year_to = 2030, vaccine_x_level = 0.7, age_from = 0, age_to = 0, gender = 3),
           non_linear_scale_up = list(year_from = 2031, year_to = 2035, endpoint = 0.9, age_from = 0, age_to = 0, gender = 3)
)
```
Then, `eval(parse())` would interpret the above set-up as 1.) there are two projection rules/functions, `catch_up_with_x` and `non_linear_scale_up`; 2.) each rule/function has a set of parameters specified as a list.  

I cannot think of a more robust way of doing this. Please let me know if you had a better approach for me to get on.

xiangli313 committed:
add reference

xiangli313 committed:
correction for at risk population

xiangli313 committed:
dplyr

xiangli313 committed:
allow multiple campaign rules

xiangli313 committed:
fix sainty check

xiangli313 committed:
typo

xiangli313 committed:
update according to using

xiangli313 committed:
updates

xiangli313 committed:
fix test

richfitz commented on README.md:
```diff
@@ -1,2 +1,5 @@
-R package for vaccine scenario generation.
+R package for vaccine scenario generation. vacsce stands for vaccination scenarios. It provides a set of coverage projection rules.
+
+## Objectives
```
this file should contain installation instructions (at a minimum, instructions for using `remotes`)

richfitz commented on vignettes/vignette.Rmd:
````diff
@@ -0,0 +1,517 @@
+
+```{r}
+# example coverage source data
+data <- read.csv("../inst/example_data.csv")
````
```suggestion
data <- system.file("example_data.csv", package = "vacsce", mustWork = TRUE)
```

richfitz commented on vignettes/vignette.Rmd:
````diff
@@ -0,0 +1,517 @@
+This is a routine vaccination coverage projection rule. It assumes that future routine coverage level is maintained at a certain constant level. In the following MCV1 example, the historical coverage of MCV1 is `r paste0(round(d$coverage[d$year == year_cur], 3)*100, "%")` in year `r year_cur`. **keep_levels** projects future coverage from `r year_cur+1` to 2030, where future coverage level is mai ...
+
+```{r}
+dat <- keep_levels(d, year_from = year_cur + 1, 
````
This won't work unless the package is loaded - you have that commented out in a hidden section above, so I suggest you put a non-hidden call to `library(vasce)` somewhere in the doc. The aim is that an interested user can follow along

richfitz commented on R/general.R:
```diff
@@ -0,0 +1,254 @@
+#' input - input parameters
+#' input is a list contains three objects - params, src and proj_rules
+#' params is a list (example below)
+#' <- list(country = "<country_code>",
```
This does not look right at all, and is causing the failing check with documentation I think. I'm not sure what you're trying to say here either so can't offer a concrete suggestion.

richfitz reviewed (changes requested):
some small comments here that should help get the checks passing - you should see a series of red 'x's on your PR that if you click through will show you issues.

You can run these locally by running `devtools::check()`

xiangli313 committed:
Update vignettes/vignette.Rmd

Co-authored-by: Rich FitzJohn <rich.fitzjohn@gmail.com>

xiangli313 committed:
Update vignettes/vignette.Rmd

Co-authored-by: Rich FitzJohn <rich.fitzjohn@gmail.com>

xiangli313 committed:
save progress

xiangli313 committed:
resolve notes and warnings

xiangli313 committed:
try a different .github

xiangli313 committed:
remove test db

xiangli313 committed:
remove vimpact

xiangli313 committed:
remove require vimpact

xiangli313 committed:
remove code, and record in inst.

xiangli313 committed:
update man, and try a different .github setup

xiangli313 committed:
upload buildegnore

xiangli313 committed:
add column to record projection rule

xiangli313 committed:
bug

xiangli313 committed:
source coverage

xiangli313 committed:
readme
